File: typescript-playwright/src/utils/generators.ts
import * as path from "path";
import { faker } from "@faker-js/faker";
import type { PhotoPayload, CommentPayload, CategoryPayload } from "./types";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const generate = {
  photoData: (overrides?: Partial<PhotoPayload>): PhotoPayload => ({
    title: `TEST-${faker.lorem.words({ min: 2, max: 4 })}`,
    author: faker.person.fullName(),
    place: faker.location.city(),
    takenAt: faker.date.past({ years: 2 }).toISOString().split("T")[0],
    ...overrides,
  }),

  commentData: (
    photoId: number,
    overrides?: Partial<CommentPayload>,
  ): CommentPayload => ({
    author: `TEST-${faker.internet.username()}`,
    text: faker.lorem.paragraph({ min: 1, max: 3 }),
    photoId,
    ...overrides,
  }),

  categoryData: (overrides?: Partial<CategoryPayload>): CategoryPayload => ({
    name: `TEST-${faker.lorem.slug(2)}`,
    ...overrides,
  }),

  ip: (): string => faker.internet.ipv4(),

  fixturePath: (filename = "test-image.jpg"): string =>
    path.resolve(__dirname, "../../src/fixtures/images", filename),
};

File: python-pytest/src/utils/generators.py
from __future__ import annotations

import random
from datetime import date
from pathlib import Path
from typing import Any, cast

from faker import Faker

from src.utils.types import CategoryPayload, CommentPayload, PhotoPayload

_fake = Faker()

_FIXTURES_IMAGES_DIR = Path(__file__).resolve().parents[2] / "src" / "fixtures" / "images"


class Generate:
    @staticmethod
    def photo_data(**overrides: Any) -> PhotoPayload:
        base: PhotoPayload = {
            "title": f"TEST-{' '.join(_fake.words(nb=random.randint(2, 4)))}",
            "author": _fake.name(),
            "place": _fake.city(),
            "takenAt": cast(
                date, _fake.date_between(start_date="-2y", end_date="today")
            ).isoformat(),
        }
        return cast(PhotoPayload, {**base, **overrides})

    @staticmethod
    def comment_data(photo_id: int, **overrides: Any) -> CommentPayload:
        base: CommentPayload = {
            "author": f"TEST-{_fake.user_name()}",
            "text": _fake.paragraph(nb_sentences=random.randint(1, 3)),
            "photoId": photo_id,
        }
        return cast(CommentPayload, {**base, **overrides})

    @staticmethod
    def category_data(**overrides: Any) -> CategoryPayload:
        base: CategoryPayload = {"name": f"TEST-{'-'.join(_fake.words(nb=2)).lower()}"}
        return cast(CategoryPayload, {**base, **overrides})

    @staticmethod
    def ip() -> str:
        return _fake.ipv4()

    @staticmethod
    def fixture_path(filename: str = "test-image.jpg") -> str:
        return str(_FIXTURES_IMAGES_DIR / filename)
